python/pyspark/sql/tvf.py

```diff
@@ -0,0 +1,714 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from typing import Optional, TYPE_CHECKING
+
+from pyspark.errors import PySparkValueError
+from pyspark.sql.dataframe import DataFrame
+from pyspark.sql.session import SparkSession
+
+if TYPE_CHECKING:
+    from pyspark.sql._typing import ColumnOrName
+
+__all__ = ["TableValuedFunction"]
+
+
+class TableValuedFunction:
+    """
+    Interface for invoking table-valued functions in Spark SQL.
+    """
+
+    def __init__(self, sparkSession: SparkSession):
+        self._sparkSession = sparkSession
+
+    def range(
```

Comment:
nit: we may make `spark.range` an alias for this new implementaion.

Reply:
Shall we do it in a separate PR? Maybe at the same time as https://github.com/apache/spark/pull/48603#discussion_r1811639985?

Reply:
sounds good!